### DISCUSSION TITLE
Installation of slideflow

### DISCUSSION THREAD

**Discussion by Nazim0001** [jamesdolezal/slideflow#190] [Q&A] [answered]
Hi,

Thank you for your work. 

I am novice in python and deep learning. Your application is usefull ;). I tried to install slideflow on terminal with the script: pip3 install slideflow as it was said from your site. But, it was not done. I see in my screen this:
![image](https://user-images.githubusercontent.com/104545136/195117921-6a93d08e-3dcf-4983-a718-74b179cdb3fd.png)

And the module slide flow is not recognized in my Jupiter Notebook.

Could you please help me to overcome this issue and use your application?

Thank you in advance for your help.
Best
Nazim

**Comment by jamesdolezal** [comment, marked as answer]
Hi Nazim, very happy to help. It looks like you are working on Windows, which is currently [not supported](https://slideflow.dev/installation.html), as I haven't had bandwidth to include Windows in our testing.

With that said, let's see if we can get it working. Are you in the default Windows environment or using [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/about) (WSL)?

**Comment by Nazim0001** [comment]
Hi,
Thank you for your prompt reply.
Yes, I am working on windows terminal and anaconda with jupiter notebook.
Thank you again for your support.
Best
Nazim

**Comment by Nazim0001** [comment]
Hi James,
Yes I am in default Windows environment. If you have some tips or solutions
please let me know.
Best
Nazim

**Comment by jamesdolezal** [comment]
The key message in the error above, `error: command 'swig.exe' failed: None`, indicates that the SWIG environment is not installed. SWIG is required to build the package `pyrfr`, which is a dependency for [SMAC](https://www.automl.org/automated-algorithm-design/algorithm-configuration/smac/), the package that provides bayesian hyperparameter optimization.

The easiest way to install SWIG on Windows is with Ananconda. The swig package is available from [conda-forge](https://anaconda.org/conda-forge/smac):

`conda install -c conda-forge smac`  

---

Although Slideflow itself is OS-agnostic, some of its dependencies may not be available on Windows, or may present challenges for installation (as with SMAC/SWIG). **There are likely to be more environmental issues beyond this when attempting to run on Windows**. Unfortunately, we don't have developer bandwidth for testing and dependency maintenance for another OS, so we will be leaving Windows as officially unsupported for now.

I would recommend switching to a Linux-based distribution (such as Ubuntu or RHEL/CentOS). If this is not possible and you choose to continue attempting to work with Windows, I would be happy to document your findings for the rest of the community if you are able to get the environment working.

**Comment by Nazim0001** [comment]
Hi James,
Thank you for your support. I will try to do at the first step on windows
environment. Just in case, do you know if slide flow could work on windows
subsystem for linux?
Thank you for your time.
Best
Nazim

**Comment by jamesdolezal** [comment]
Yes - Slideflow will work using WSL2, which uses a Ubuntu environment (just tested), with the exception of Workbench as WSL does not support graphical applications. Follow the installation instructions for Ubuntu and you should be good to go.

**Comment by Nazim0001** [comment]
Hi James,

I finally installed the Ubuntu desktop because WSL2 doesn't work. I
followed the tutorial of installation of the site.

These 3 lines did not work:
$ python setup.py bdist_wheel
$ pip install dist/slideflow-1.X.X-py3-any.whl

$ docker run -it --shm-size=2g --gpus all
jamesdolezal/slideflow:latest-torch

After that I thought that it was not important, so I ran python3 and wrote:
'import slideflow as sf'... but it doesn't work.

When I ran 'pip install slideflow', this error was written:
Collecting heapdict
  Using cached HeapDict-1.0.1-py3-none-any.whl (3.9 kB)
Building wheels for collected packages: imgui
  Building wheel for imgui (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for imgui (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [45 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-cpython-310
      creating build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/extra.py -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/__init__.py -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/_compat.py -> build/lib.linux-x86_64-cpython-310/imgui
      creating build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/pygame.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/base.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/cocos2d.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/pyglet.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/sdl2.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/glfw.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/__init__.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/opengl.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      copying imgui/integrations/glumpy.py ->
build/lib.linux-x86_64-cpython-310/imgui/integrations
      running egg_info
      writing imgui.egg-info/PKG-INFO
      writing dependency_links to imgui.egg-info/dependency_links.txt
      writing requirements to imgui.egg-info/requires.txt
      writing top-level names to imgui.egg-info/top_level.txt
      reading manifest file 'imgui.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      no previously-included directories found matching 'imgui-cpp/examples'
      no previously-included directories found matching
'imgui-cpp/extra_fonts'
      adding license file 'LICENSE'
      writing manifest file 'imgui.egg-info/SOURCES.txt'
      copying imgui/ansifeed.pxd -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/cimgui.pxd -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/core.cpp -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/core.h -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/core.pxd -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/core.pyx -> build/lib.linux-x86_64-cpython-310/imgui
      copying imgui/enums.pxd -> build/lib.linux-x86_64-cpython-310/imgui
      running build_ext
      building 'imgui.core' extension
      creating build/temp.linux-x86_64-cpython-310
      creating build/temp.linux-x86_64-cpython-310/config-cpp
      creating build/temp.linux-x86_64-cpython-310/imgui-cpp
      creating build/temp.linux-x86_64-cpython-310/imgui
      gcc -pthread -B /home/nazim/anaconda3/envs/slideflow/compiler_compat
-Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2
-isystem /home/nazim/anaconda3/envs/slideflow/include -fPIC -O2 -isystem
/home/nazim/anaconda3/envs/slideflow/include -fPIC
-DPYIMGUI_CUSTOM_EXCEPTION -Iimgui-cpp -Iansifeed-cpp -I./imgui -Iimgui
-Iconfig-cpp -I/home/nazim/anaconda3/envs/slideflow/include/python3.10 -c
config-cpp/py_imconfig.cpp -o
build/temp.linux-x86_64-cpython-310/config-cpp/py_imconfig.o
-includeconfig-cpp/py_imconfig.h
      error: command 'gcc' failed: No such file or directory
      [end of output]

  note: This error originates from a subprocess, and is likely not a
problem with pip.
  ERROR: Failed building wheel for imgui
Failed to build imgui
ERROR: Could not build wheels for imgui, which is required to install
pyproject.toml-based projects

So, I tried to overcome the issue with installation of pyproject.toml with
pip, but the installation of slideflow gave me the error.

I don't understand why slideflow doesn't work. Could you please help for
that?

Thank you in advance.
Best
Nazim


*---*

*Nazim BENZERDJEB, MD, PhD *

PHU

Pathology Department

University Hospital of Lyon - South

165 Chemin du Grand Revoyet

69495 Pierre Bénite cedex

France

**Comment by jamesdolezal** [comment]
The error message comes from a failure to install the dependency imgui:
```
  ERROR: Failed building wheel for imgui
```
Which occurs because the installer can't find gcc:
```
error: command 'gcc' failed: No such file or directory
```

If this is a fresh installation of Ubuntu, you may need to first install the GCC compilation tools, with:

```
sudo apt update
sudo apt install build-essential
```

Some follow-up questions:

- What version of Ubuntu are you using?
- What environment are you using (Python virtualenv, Anaconda, etc?)
- What Python version are you using?
- Are you using PyTorch or Tensorflow, and which version are you using?

Also, for clarification - you can install **either** with PIP _or_ Docker. You don't need to do both. The instructions and questions above are assuming you're installing with PIP (which it looks like you are). If this is the case, ignore the Docker installation instructions as they won't apply.

**Comment by Nazim0001** [comment]
Thank you James. The installation of slideflow is good. But, when I tried
this:

(slideflow) ***@***.***:~/slideflow$ python3 run_project.py

Traceback (most recent call last):
  File "/home/nazim/slideflow/run_project.py", line 39, in <module>
    import slideflow as sf
  File "/home/nazim/slideflow/slideflow/__init__.py", line 36, in <module>
    from slideflow import io, model, norm, stats
  File "/home/nazim/slideflow/slideflow/io/__init__.py", line 20, in
<module>
    from slideflow.io.tensorflow import get_tfrecord_parser  # noqa F401
  File "/home/nazim/slideflow/slideflow/io/tensorflow.py", line 15, in
<module>
    from slideflow.io import gaussian
  File "/home/nazim/slideflow/slideflow/io/gaussian.py", line 3, in <module>
    import tensorflow as tf
ModuleNotFoundError: No module named 'tensorflow'

I don't understand, because slideflow was perfectly installed.

For your questions:

   - What version of Ubuntu are you using? *Ubuntu 22.04 LTS*
   - What environment are you using (Python virtualenv, Anaconda, etc?)
   *anaconda*
   - What Python version are you using? *python 3*
   - Are you using PyTorch or Tensorflow, and which version are you using? *I
   want to use PyTorsh (1.12.1)*

Maybe I lacked something.

Thank you for your help.
Best
Nazim

**Comment by jamesdolezal** [comment]
Slideflow requires **either** Tensorflow or PyTorch, as listed in the [requirements](https://slideflow.dev/installation.html#software-requirements) section of the installation instructions. The PyTorch installation instructions for a conda environment can be found [here](https://pytorch.org/get-started/locally/).

After PyTorch is installed, you'll need to install the necessary PyTorch-related packages (torchvision, pretrained_models), which can be done with:

```
pip3 install slideflow[torch]
```

By default, Slideflow will expect to use the Tensorflow backend. To switch to the PyTorch backend, after PyTorch is installed (by following the above link), set the environmental variable `SF_BACKEND` equal to `torch`:

```
export SF_BACKEND=torch
```

as described in the [Changing backends](https://slideflow.dev/installation.html#changing-backends) section of the installation instructions.

You can either run this command each time you open a new terminal, or you can permanently set the environmental variable by adding that line to your `.bashrc` file in your home folder (see [this page](https://askubuntu.com/questions/58814/how-do-i-add-environment-variables) for more information on setting an environmental variable)

**Comment by Nazim0001** [comment]
Hi James,
I installed slideflow and Pytorch,...put GPU. But, the import slideflow in
python does not work and I read that:

(slideflow) ***@***.***:~/slideflow$ python3
Python 3.10.6 (main, Oct  7 2022, 20:19:58) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import slideflow








*Traceback (most recent call last):  File "<stdin>", line 1, in <module>
File "/home/nazim/slideflow/slideflow/__init__.py", line 36, in <module>
from slideflow import io, model, norm, stats  File
"/home/nazim/slideflow/slideflow/model/__init__.py", line 39, in <module>
  from slideflow.model.torch import (CPHTrainer, Features, load, # noqa
F401  File "/home/nazim/slideflow/slideflow/model/torch.py", line 13, in
<module>    import pretrainedmodelsModuleNotFoundError: No module named
'pretrainedmodels'*
>>>

Please, do you have any solution for this?
best regards
Nazim

*---*

*Nazim BENZERDJEB, MD, PhD *

PHU

Pathology Department

University Hospital of Lyon - South

165 Chemin du Grand Revoyet

69495 Pierre Bénite cedex

France

**Comment by jamesdolezal** [comment]
Hello Nazim,

The error `No module named 'pretrainedmodels'` indicates that the package 'pretrainedmodels' is not installed. This package should have been installed if you ran the command:

```
pip3 install slideflow[torch]
```

If not, you can manually install the package with:

```
pip3 install pretrainedmodels
```

If you encounter any similar issues with ModuleNotFound errors, you can try the same approach using `pip3 install` to install the missing packages.

**Comment by Nazim0001** [comment]
Thank you for your response. It is working now 😁

I don t want to bother you, but do you have a solution to run
jupiter-notebook?

Best
Nazim

**Comment by jamesdolezal** [comment]
Slideflow works in Jupyter notebooks without any extra configuration. You can use it as you would in a script.

You can find instructions for installing and using jupyter notebooks on their website:  https://jupyter.org/install

And there is some good documentation for how to use notebooks here: https://docs.jupyter.org/en/latest/start/index.html

**Comment by Nazim0001** [comment]
Hi James,
thank you so much, it is working with Jupyter Notebook as well.
Best

*---*

*Nazim BENZERDJEB, MD, PhD *

PHU

Pathology Department

University Hospital of Lyon - South

165 Chemin du Grand Revoyet

69495 Pierre Bénite cedex

France

**Comment by Nazim0001** [comment]
Hi james,

I have a couple of questions:

1- after installation of slideflow on pip and Jupiter, I tried to run it
through Jupyter-notebook.
import slideflow as sf
P = sf.Project(
    '/home/nazim/Documents/project/Gastric',
    name="Gastric",

annotations="/home/nazim/Documents/project/Gastric/annotation_final.csv")

When I run the above script, it doesn't work on Jupyper-notebook:

AttributeError                            Traceback (most recent call
last)/tmp/ipykernel_65316/478529060.py in <module>      1 import
slideflow as sf----> 2 P = sf.Project(      3
'/home/nazim/Documents/project/Gastric',      4     name="Gastric",
  5     annotations="/home/nazim/Documents/project/Gastric/annotation_final.csv")


AttributeError: module 'slideflow' has no attribute 'Project'


Do you have please any idea of a solution to this issue?


2- About the meaning behind items, I don't understand what is the
difference for you in terms of :

- annotation.csv (list of name of cases and what do you expect in this file
: name of file, categories (HVP +, HPV -, ....),

- for roi, just jason files


3- for slides and its directory : what do you expect ? directories of files
of slides ?


Thank you for your replies.

Have a good day

best

*---*


*Nazim*

**Comment by jamesdolezal** [comment]
1. `AttributeError: module 'slideflow' has no attribute 'Project'`

This is most likely caused by a folder in your current working directory with the name "slideflow". When you import a module, the first thing python will do is look for a folder in your current directory with that name. If you cloned the github repository and there is a "slideflow" folder where you are running the code, it will try to import the github repository folder, which is not a valid module, hence the error.

You can either:

- Rename the cloned github repository folder from "slideflow" to something else
- Run your jupyter notebook in a different folder
- Enter the github "slideflow" folder and run from inside the folder

You can find more information about python imports here: https://realpython.com/absolute-vs-relative-python-imports/

2. The "annotations" file is a file containing patient-level outcome categories, such as HPV status for each patient/slide. You can find more information along with a specific example [in the documentation](https://slideflow.dev/project_setup.html#setting-up-annotations). You might also find the [first tutorial](https://slideflow.dev/tutorial1.html) helpful. 

3. The slides directory, when you are setting up a dataset, should be a path to where all the slides (e.g. *.svs files) are saved. More information on setting up datasets can be found [here](https://slideflow.dev/project_setup.html#configuring-datasets).

**Comment by Nazim0001** [comment]
Hi James,

Thank you for your reply. It is helpful.

I would like to see the extraction of tiles but I read this error after
this "dataset.extract_tiles()":

LiveError                                 Traceback (most recent call last)
Input In [11], in <cell line: 1>()----> 1 dataset.extract_tiles()

File ~/anaconda3/envs/slideflow/lib/python3.10/site-packages/slideflow/dataset.py:1186,
in Dataset.extract_tiles(self, save_tiles, save_tfrecords, source,
stride_div, enable_downsample, roi_method, skip_extracted, tma,
randomize_origin, buffer, q_size, qc, report, **kwargs)   1180
speed_task = pb.add_task("Speed: ",   1181
progress_type="speed",   1182                          total=None)
1183 slide_task = pb.add_task("Extracting...",   1184
        progress_type="slide_progress",   1185
 total=len(slide_list))-> 1186 pb.start()   1187 if buffer:   1188
# Worker to put each slide path into queue   1189     def
buffer_worker():

File ~/anaconda3/envs/slideflow/lib/python3.10/site-packages/rich/progress.py:1160,
in Progress.start(self)   1158 """Start the progress display."""
1159 if not self.disable:-> 1160     self.live.start(refresh=True)

File ~/anaconda3/envs/slideflow/lib/python3.10/site-packages/rich/live.py:113,
in Live.start(self, refresh)    111 if self._started:    112
return--> 113 self.console.set_live(self)    114 self._started = True
  115 if self._screen:

File ~/anaconda3/envs/slideflow/lib/python3.10/site-packages/rich/console.py:840,
in Console.set_live(self, live)    838 with self._lock:    839     if
self._live is not None:--> 840         raise errors.LiveError("Only
one live display may be active at once")    841     self._live = live
LiveError: Only one live display may be active at once


Do you have any idea of this issue?


Thank you in advance for your assistance.

Best

Nazim

**Comment by jamesdolezal** [comment]
This is probably because execution of a cell was stopped/interrupted, which caused the progress bar to improperly shut down.

I can provide more details and help you troubleshoot, but let's have you open an Issue Report about this first (https://github.com/jamesdolezal/slideflow/issues). Create an issue, fill out the form, and we'll continue the discussion there. It's easier for tracking individual problems, and allows us to reference previous discussions if someone else has the same problem in the future.
